### PR TITLE
automate ui testing

### PR DESCRIPTION
**Pull request by yann300**
address https://github.com/ethereum/browser-solidity/issues/280

### CODE REVIEW

File: test-browser/tests/ballot.js [hunk 1]
```diff
@@ -1,35 +1,27 @@
...
 
 function runTests (browser, testData) {
   browser
     .url('http://127.0.0.1:8080/#version=builtin')
     .waitForElementVisible('.newFile', 10000)
-  browser.assert.notEqual(testData, null)
-// TODO add Ballot tests. -> setValue('#input textarea', ... ) is not working properly with that contract.
-/*
-  testBallot(browser, testData.ballot.sources.Untitled, function () {
+  contractHelper.testContracts(browser, sources.sources.Untitled.replace(/(\n)/g, ' '), ['Ballot'], function () {
```
Comment (chriseth) [inline test-browser/tests/ballot.js]: Why do you need the replacement?

Comment (chriseth) [inline test-browser/tests/ballot.js]: Oh wow, but that has to be fixed somewhere else - we cannot use this workaround in every single test...

Comment (yann300) [inline test-browser/tests/ballot.js]: i will move it to 
https://github.com/ethereum/browser-solidity/pull/283/files#diff-f7a6123b8a31b64951b350d64289b3abR24

**Comment by chriseth** [comment]
Did you try `editor.setBehavioursEnabled(false)` to disable the auto-closing of parentheses?

**Comment by yann300** [comment]
yes but i had hardtime to retrieve a ref of the editor from the browser.execute(..) of nightwatch

**Comment by yann300** [comment]
it might be possible to have a global flag (test mode) that tell the editor to turn off autoclose bracket. 
But I don't like to put the test if (flag) {} in the editor.js file

File: ci/makeMockCompiler.js [hunk 2]
```diff
@@ -4,20 +4,39 @@ var fs = require('fs')
...
+})
+
+function gatherCompilationResults (callback) {
+  var compilationResult = {}
+  fs.readdir('./test-browser/tests', 'utf8', function (error, data) {
+    if (error) {
+      console.log(error)
+      process.exit(1)
+    } else {
+      data.map(function (item, i) {
```
Comment (chriseth) [inline ci/makeMockCompiler.js]: Perhaps rename `data` to `filenames`?

File: ci/makeMockCompiler.js [hunk 3]
```diff
@@ -4,20 +4,39 @@ var fs = require('fs')
...
+        var testDef = require('../test-browser/tests/' + item)
+        for (var k in testDef) {
```
Comment (chriseth) [inline ci/makeMockCompiler.js]: Why not just use `if ('@Sources' in testDef)`? Also, could we change that to `@sources`?

File: src/app/editor.js [hunk 4]
```diff
@@ -12,6 +12,7 @@ function Editor (loadingFromGist, storage) {
   var SOL_CACHE_FILE = null
 
   var editor = ace.edit('input')
+  document.getElementById('input').editor = editor
```
Comment (chriseth) [inline src/app/editor.js]: Please add a comment: `required to access the editor during tests`

File: test-browser/helpers/applytestmode.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+/**
+ * This script is injected by NightWatch just before starting test
+ *
+ */
+console.log('applying test mode')
+document.getElementById('input').editor.setBehavioursEnabled(false)
```
Comment (chriseth) [inline test-browser/helpers/applytestmode.js]: Please add comment: disable bracket auto-match (i.e. automatic injection of closing brackets and other things), so we can enter raw source code.

File: test-browser/helpers/contracts.js [hunk 6]
```diff
@@ -1,25 +1,27 @@
...
-      browser.assert.equal(result.value.length, compiled.length)
-      result.value.map(function (item, i) {
-        browser.assert.equal(item.name, compiled[i])
+  checkCompiledContracts: checkCompiledContracts,
+  testContracts: testContracts
+}
+
+function checkCompiledContracts (browser, compiled, callback) {
+  browser.elements('css selector', '.udapp .contract .title', function (elements) {
+    console.log(JSON.stringify(elements))
```
Comment (chriseth) [inline test-browser/helpers/contracts.js]: Is this needed? (or even visible?)

File: test-browser/helpers/contracts.js [hunk 7]
```diff
@@ -1,25 +1,27 @@
...
+    callback()
+  })
+}
+
+function testContracts (browser, contractCode, compiledContractNames, callback) {
+  browser
+      .clearValue('#input textarea')
+      .click('.newFile')
+      .setValue('#input textarea', contractCode, function () {})
+      .waitForElementPresent('.contract .create', 3000000)
```
Comment (chriseth) [inline test-browser/helpers/contracts.js]: I think we can now reduce the timeout.

File: test-browser/tests/ballot.js [hunk 8]
```diff
@@ -1,35 +1,30 @@
...
   'Ballot': function (browser) {
-    runTests(browser, testData)
+    runTests(browser)
   }
 }
 
 function runTests (browser, testData) {
   browser
-    .url('http://127.0.0.1:8080/#version=builtin')
     .waitForElementVisible('.newFile', 10000)
```
Comment (chriseth) [inline test-browser/tests/ballot.js]: Do we have to provide an explicit timeout?

Comment (yann300) [inline test-browser/tests/ballot.js]: the second param is not optional

Comment (chriseth) [inline test-browser/tests/ballot.js]: oh right, I'm sorry

**Pull request merged**